### PR TITLE
[LLVM][TableGen] Change all `Init` pointers to const

### PR DESCRIPTION
**Pull request by jurahul**
This is a part of effort to have better const correctness in TableGen backends:

https://discourse.llvm.org/t/psa-planned-changes-to-tablegen-getallderiveddefinitions-api-potential-downstream-breakages/81089

### CODE REVIEW

**Comment by topperc** [comment]
Do Init objects have any methods that can modify them? They const was explicitly removed from some uses a long time ago af8ee2cdeeb2ed35fdf0db8f7c06254c5d333ac4.

**Comment by jurahul** [comment]
> Do Init objects have any methods that can modify them? `const` was explicitly removed from some uses a long time ago [af8ee2c](https://github.com/llvm/llvm-project/commit/af8ee2cdeeb2ed35fdf0db8f7c06254c5d333ac4).
> 
> Here's a very old thread about removing `const` from IR `Type` pointers because they are immutable. https://lists.llvm.org/pipermail/llvm-commits/Week-of-Mon-20110711/124192.html I assume a similar thinking was being applied to Init objects.

That's a good point. Though I do see "const Type*" pointers in the code. I'd think there is a difference between an implementation detail of "this class is uniqued in the impl and therefore immutable" vs "irrespective of being uniqued or not, in this code I do not intend to modify it". The const-ness of pointers expresses the second one, regardless of the implementation detail of whether the object itself is immutable or not. But I'll let the reviewers decide. Note that some changes of this nature have been committed earlier for Clang and MLIR tablegen backends (as a prep for this one).  Also, this enables us to get rid of a bunch of const_casts<> in the various Init::Fold() methods (which are const, but currently return non-const Init ptr, and when the Fold is unresolved have to const_cast<> this to non-const pointer to return it). So making them const just seems to play nicer with C++.

**Comment by topperc** [comment]
> Also, this enabled removing one const_cast<> in #112261 (which I think is going to be the final change for the time being to make more pointers const in TableGen backends).

That one seems related to Record pointers being const. Can that be fixed independent of Init pointers being const?

**Comment by arsenm** [comment]
I think const should always be used in all situations

File: llvm/lib/TableGen/Record.cpp [hunk 1]
```diff
@@ -551,18 +552,18 @@ std::string BitsInit::getAsString() const {
...
+  const Init *CachedBitVarResolved = nullptr;
 
   for (unsigned i = 0, e = getNumBits(); i != e; ++i) {
-    Init *CurBit = getBit(i);
-    Init *NewBit = CurBit;
+    const Init *CurBit = getBit(i);
+    const Init *NewBit = CurBit;
 
-    if (VarBitInit *CurBitVar = dyn_cast<VarBitInit>(CurBit)) {
+    if (const VarBitInit *CurBitVar = dyn_cast<VarBitInit>(CurBit)) {
```
Comment (topperc) [inline llvm/lib/TableGen/Record.cpp:566]: Could use `auto *CurBitVar` here and on all the other dyn_cast/cast. But you don't have to.

**Review by arsenm — APPROVED** [review]

File: llvm/lib/TableGen/Record.cpp (hunk 1 above)
Comment (jurahul) [reply llvm/lib/TableGen/Record.cpp:566]: I will do this in a follow up PR.

**Pull request merged**